### ISSUE TITLE
Make RuntimeTrainer.image optional when image is provided via CustomTrainer / CustomTrainerContainer

### ISSUE THREAD

**Issue by kliushenkov** [kubeflow/sdk#494]
### What you would like to be added?

Treat the container image **as optional on RuntimeTrainer**, because callers can supply it on the job trainer instead of requiring it on every `TrainingRuntime` / `ClusterTrainingRuntime`.
**Type change**
```python
@dataclass
class RuntimeTrainer:
    trainer_type: TrainerType
    framework: str
    image: str | None = None  # was: image: str
    num_nodes: int = 1
    device: str = common_constants.UNKNOWN
    device_count: str = common_constants.UNKNOWN
    ...
```

**Behavior**
1. get_runtime_trainer() ([kubeflow/trainer/backends/kubernetes/utils.py](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/backends/kubernetes/utils.py#L69))
- Set `image=trainer_container.image` when the runtime CR defines one.
- Allow `image=None` when the runtime template has no image (minimal runtime: framework, resources, command only).
2. TrainJob creation (`get_trainer_cr_from_custom_trainer` / similar)
- Resolve image in order: `CustomTrainer.image` / `CustomTrainerContainer.image` → `runtime.trainer.image` → error if still missing.
-  Fix validation to actually `raise ValueError(...)` when no image is available from any source.
3. Backends
- Kubernetes: only set `trainer_cr.image` when resolved.
-  Container / localprocess: handle `runtime.trainer.image is None` when `CustomTrainer` / `CustomTrainerContainer` provides `image` (container backend already has` trainer.image if trainer.image else runtime.trainer.image` in places — align the rest).
4. Tests & docs
- Runtime parsed without image + `CustomTrainer(image="...")` succeeds.
- Runtime without image + trainer without image fails with a clear error.
- `BuiltinTrainer` paths unchanged where runtime image is still required.

**Example**
```python
# Runtime CR: no image in template — only framework/resources
client.train(
    runtime="minimal-torch-runtime",
    trainer=CustomTrainer(
        func=train_fn,
        image="my-registry.io/train:v1",
    ),
)

# Or explicit container trainer
client.train(
    runtime="minimal-torch-runtime",
    trainer=CustomTrainerContainer(image="my-registry.io/train:v1"),
)
```

### Why is this needed?

**Redundant and inflexible modeling**

`CustomTrainer` and `CustomTrainerContainer` already support an optional/required `image` on the **job** trainer. Forcing `RuntimeTrainer.image: str` implies every runtime CR must bake in an image, even when teams always override it per job (notebooks, CI, experiments).

**Lighter `TrainingRuntime` definitions**

A runtime can describe **how** to train (framework, nodes, devices, default command) while the what to run image lives on the trainer. That matches common patterns:

- shared cluster `ClusterTrainingRuntime` without a pinned image;
- namespaced `TrainingRuntime` for team defaults;
- per-run image on `CustomTrainer` / `CustomTrainerContainer`.

**Avoid duplicate configuration**

Requiring runtime image + trainer image invites drift (runtime says one image, job uses another). Optional runtime image makes the trainer the single source of truth when provided.

**Correct validation semantics**

Image should be required for a **submitted TrainJob**, not for every **listed Runtime** object. Validation belongs at job build time (“runtime image OR trainer image must be set”), not when parsing a runtime that is only used for metadata until a trainer is attached.

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

**Comment by andreyvelich** [comment]
Thanks for creating this @kliushenkov!
I think, this is valid request. However, creation of a Runtime without an image will also fail this API: `get_runtime_packages()`.

Could you share more about your use-case when you don't want to specify a base image in the Runtime configuration?
So users are fully responsible to bring their own images.

/remove-label needs-triage

cc @kubeflow/kubeflow-sdk-team

**Comment by kliushenkov** [comment]
Hi @andreyvelich !
Thanks for the review and for flagging get_runtime_packages()
> Could you share more about your use-case when you don't want to specify a base image in the Runtime configuration? So users are fully responsible to bring their own images.

In our setup, each templated repo owns its training environment end-to-end (independent uv env + Docker image). Examples:
| Model    | Stack (simplified) |
| -------- | ------- |
| ALS  | Python 3.10, CUDA   |
| SASRec | Python 3.10, PyTorch, CUDA    |
| Boosting ranker    | Python 3.10   |
| EASE    | Python 3.12, NumPy 2+   |
	
We do not want a single “platform” training image baked into `ClusterTrainingRuntime`, and we also do not want one `ClusterTrainingRuntime` (or `TrainingRuntime`) **per model**. In our case that would mean many nearly identical CRs that differ only by `image` (ALS, SASRec, EASE, boosting-ranker, etc.), which duplicates MLOps config and makes every image tag change a runtime CR update. Instead, users bring their own image on each job via `CustomTrainer(image=...)` or `CustomTrainerContainer(image=...)`, while a small set of shared runtimes (e.g. CPU vs GPU) holds only infrastructure settings.

What we do want in `ClusterTrainingRuntime` / `TrainingRuntime` (owned by MLOps) is infrastructure policy, not application dependencies:
- annotations, env, affinity, tolerations
- separation of CPU vs GPU cluster runtimes (e.g. one for EASE/boosting, one for ALS/SASRec)

So the runtime describes **how/where** the job runs on the cluster; the trainer describes what runs (image + code).

That matches the split already implied by `CustomTrainer.image` / `CustomTrainerContainer.image` — we’d just like `RuntimeTrainer.image` to be optional so listing/parsing a runtime doesn’t require a placeholder image we never use.

We also rotate **Docker image tags per repository** when testing new features (branch builds, RCs). The tag should be passed on each job via `CustomTrainer(image=...)` or `CustomTrainerContainer(image=...)`, not updated on `ClusterTrainingRuntime` / `TrainingRuntime`. The runtime stays fixed for infra; only the trainer image changes between experiments.

Typical flow:
```python
client.train(
    runtime="cpu-cluster-runtime",
    trainer=CustomTrainerContainer(
        image="registry.example.com/ease:numpy2-experiment-7",
    ),
)
```

> However, creation of a Runtime without an image will also fail this API: `get_runtime_packages()`.

You’re right: today this API **depends on `runtime.trainer.image`**:
- Container backend: runs a one-off container with `runtime.trainer.image`.
- Kubernetes backend: submits a `CustomTrainer` job without passing `trainer.image`, so it relies on the image from the `TrainingRuntime` / `ClusterTrainingRuntime` CR.

For image-less runtimes + BYOI, we don’t rely on `get_runtime_packages()` in normal workflows — we already know what’s inside images we build in CI. So making `RuntimeTrainer.image` optional for `train()` shouldn’t block our main path.

For the SDK, I'd suggest one of these (happy to help with a PR):
- Optional `image` argument (good for BYOI):
`get_runtime_packages(runtime, image="registry.example.com/als:feature-xyz")`
so users can inspect their own image while keeping infra-only runtimes.
- Optional `trainer` argument (aligns with `train()`):
`get_runtime_packages(runtime, trainer=CustomTrainer(..., image="..."))`
or `trainer=CustomTrainerContainer(image="...")`
and resolve the image the same way as job submission: `trainer.image` → `runtime.trainer.image` → error if still missing.
This avoids duplicating the image string when the user already has a trainer object for experiments.

**Comment by andreyvelich** [comment]
Yeah, I think this is fair design, thanks for sharing this @kliushenkov!

Let's set `image` to Unknown by default in the Runtime, like we do [for devices](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/types/types.py#L255).

For `get_runtime_package()` API we can just Raise exception if image is unset, like [we do for BuiltinTrainer](https://github.com/kubeflow/sdk/blob/main/kubeflow/trainer/backends/kubernetes/backend.py#L226-L227).

@kliushenkov Would you be interested to contribute this?

**Comment by kliushenkov** [comment]
Hi @andreyvelich — thanks, that sounds good to us.

Yes, I’d like to contribute. I’ll follow the approach we discussed (`Unknown` default, `get_runtime_packages()` error when unset, image resolution in `train()`)

**Comment by OadeO6** [comment]
Hi @kliushenkov, You might want to close issue #468 since it's more or less the same as this one which you're already working on, to avoid duplicates.
